### PR TITLE
initial commit of Get-DbaMemoryUsage

### PR DESCRIPTION
**Pull request by PowerDBAKlaas**
Has been tested on minimum requirements:
- [ ]  PowerShell 3
- [ ]  Windows 7
- [ ]  SQL Server 2000

Has been tested on maximum requirements:
- [ ]  SQL Server vNext
- [ ]  Windows 10
- [ ]  Azure Database

Tests for tester:
- [ ] Working/useful help content, including link to command on dbatools web site
- [ ] All examples work as advertised
- [ ] Does not contain template content
- [ ] Does not contain excessive/unnecessary amounts of comments
- [ ] Works remotely
- [ ] Works locally
- [ ] Works on lower versions or throws error specifying version not supported
- [ ] Works with named instances
- [ ] Works with clustered instances
- [ ] Handles offline/read only databases
- [ ] Supports multiple servers (at the command line or piped from Get-SqlRegisteredServerName)
- [ ] No un-handled errors which stop the command working with multiple servers

### CODE REVIEW

File: functions/Get-DbaMemoryUsage.ps1 [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+    }
+
+	PROCESS
+	{
+        foreach ($servername in $ComputerName)
+        {
+            if ($servername -match '\\')
+			{
+				$servername = $servername.Split('\\')[0]
```
Comment (ctrlbold) [inline functions/Get-DbaMemoryUsage.ps1]: They may throw an aliased server name. Please use the newly created internal command Resolve-IpAddress to get the IP. You can throw a few things at it: a string, an smo object or even a string with a named instance.

File: functions/Get-DbaMemoryUsage.ps1 [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+			}
+            Write-Verbose "Connecting to $servername"
+			if ( test-connection -ComputerName $servername -Quiet -count 1)
+            {
+                $availablecounters = (get-counter -ComputerName $servername -ListSet '*sql*:Memory Manager*').paths
```
Comment (ctrlbold) [inline functions/Get-DbaMemoryUsage.ps1]: Please use upper case for all commands, including Test-Connection

File: functions/Get-DbaMemoryUsage.ps1 [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+				                MemKB = $_.cookedvalue
+				                MemMB = $_.cookedvalue / 1024
+                                }
+                            }
+
+
+                $availablecounters = (get-counter -ComputerName $servername -ListSet '*sql*:Plan Cache*' ).paths
+                (get-counter -ComputerName $servername -Counter $availablecounters).countersamples |
+                    Where-Object {$_.Path -match $Plancounters} |
+                    foreach { [PSCustomObject]@{
```
Comment (ctrlbold) [inline functions/Get-DbaMemoryUsage.ps1]: love 💃

File: functions/Get-DbaMemoryUsage.ps1 [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            }
+			else
+			{
+				Write-Warning "Can't connect to $servername. Moving on."
+				Continue
+			}
+        }
+    }
+    END
+    {}
```
Comment (ctrlbold) [inline functions/Get-DbaMemoryUsage.ps1]: You can remove this part

File: functions/Get-DbaMemoryUsage.ps1 [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+                                Counter = $_.Path.split("\")[-1]
+					            Pages = $_.cookedvalue
+					            MemKB = $_.cookedvalue * 8192 / 1024.0
+					            MemMB = $_.cookedvalue * 8192 /1048576.0
+                                }
+                            }
+            }
+			else
+			{
+				Write-Warning "Can't connect to $servername. Moving on."
```
Comment (ctrlbold) [inline functions/Get-DbaMemoryUsage.ps1]: Is there a way for us to catch this earlier?

Comment (PowerDBAKlaas) [inline functions/Get-DbaMemoryUsage.ps1]: I don't think so: the test-connection is inside the if clause. If this returns nothing, the if {} block is skipped and the next step is the else {} block.

File: functions/Get-DbaMemoryUsage.ps1 (hunk 2 above)
Comment (PowerDBAKlaas) [inline functions/Get-DbaMemoryUsage.ps1]: done

File: functions/Get-DbaMemoryUsage.ps1 (hunk 4 above)
Comment (PowerDBAKlaas) [inline functions/Get-DbaMemoryUsage.ps1]: it's neat to leave it in, according to me and Don Jones :-)

**Comment by ctrlbold** [comment]
Hey @PowerDBAKlaas, can you make the change to the IP/Hostname and I can test and merge? So many PR's have been stuck.

**Comment by ctrlbold** [comment]
Also I'm not sure why but there are too many files in this PR, including psd1

**Comment by PowerDBAKlaas** [comment]
I've seen those files, and I don't know what they have to do with this PR, or how they ended up in my branch. Can a Github specialist fix this?

How should I handle the hostname issue right now? use resolve-ipaddress? or Get-Connection? I haven't come to write a better version of the internal function yet.

**Comment by ConstantineK** [comment]
Sorry that I lost some of your commits, but it looked like the files had been changed in a previous version and was having trouble merging, so I did the following:

- Renamed the existing branch (git branch -m old_branch new_branch).
- Created a new branch from development (git checkout development; git checkout -b New-Get-DbaMemoryUsage), remember that case counts!
- Checked out the code we were interested in (git checkout old_branch filepath) to add it to our existing branch.
- Committed and force pushed the code to this branch (git commit -m "message"; git push origin New-Get-Dba-MemoryUsage --force)

**Comment by ctrlbold** [comment]
oh man, that 1 files changed is gorgeous! amazing work @ConstantineK !

**Comment by ctrlbold** [comment]
This returns memory usage for all instances on a server. When using it, I'd assume it would be per SQL Server.

**Comment by PowerDBAKlaas** [comment]
@ctrlbold using Resolve-DbaNetworkName for the first time. 😃

**Comment by PowerDBAKlaas** [comment]
@ctrlbold can you show the output you get, because I get a counter per sqlinstance (as long as the counters are present on the computer, they can easily go missing 😢 )

**Comment by ctrlbold** [comment]
Yes, here, sorry :(

image

**Comment by PowerDBAKlaas** [comment]
and what would you like to be different?

**Comment by ctrlbold** [comment]
My initial impression is that it was for _one_ SQL Server instance. So the output should only be related to the SQL Instance that was passed. Otherwise, this should be a ComputerName only parameter. 

Even if that is the case, the output should be parsed and an extra column should be added that has the instance name. In this case, sqlserver seems like MSSQLSERVER/default and mssql$instancename seems like the named instances.

Ultimately, it'd be awesome to have -- if they pass ComputerName, then all instances, but if they pass SqlInstance, then the specific instance.

**Comment by PowerDBAKlaas** [comment]
it is ComputerName, but
as you asked to "handle anything they can throw at us", I accept sqlinstances and get the computername out of it, so even 'realcomputer\nonexistingInstance' will work.
All this is handled by the praised Resolve-DbaNetworkName. 😃 
I removed the 'sqlinstance' and 'sqlserver' aliases.

I put the sqlinstance in a separate property.

This is not the ultimate version yet, but acceptable as an early release?

**Pull request merged**